### DISCUSSION TITLE
动态库的十分奇怪的问题...

### DISCUSSION THREAD

**Discussion by Extra-Creativity** [xmake-io/xmake#4521] [Q&A] [answered]
非常简单的一个测试（已附在附件里）：
![image](https://github.com/xmake-io/xmake/assets/72960952/bee0092a-6235-4b1f-b62b-12079ee36dc5)
以下是`xmake.lua`：
```lua
set_languages("cxx17")
add_rules("mode.debug", "mode.release")

target("foo")
    set_kind("shared")
    add_files("src/func.cpp")

target("test")
    set_kind("binary")
    add_deps("foo")
    add_files("src/main.cpp")
```
但是出现报错：
![image](https://github.com/xmake-io/xmake/assets/72960952/85ed86f6-ee7f-4e06-873b-23296d268d94)
我强行用msvc的命令链接，又出现报错`fatal error LNK1107: 文件无效或损坏: 无法在 0x2F8 处读取`；所以为什么`add_deps`会对动态库尝试`.lib`呢？生成的好像也有点问题？
环境是Windows10/msvc19.29（即VS2019最新版本）
[test2.zip](https://github.com/xmake-io/xmake/files/13715745/test2.zip)

**Comment by star-hengxing** [comment, marked as answer]
https://xmake.io/#/zh-cn/manual/custom_rule?id=utilssymbolsexport_all

**Reply by Extra-Creativity** [reply]
我读文档不仔细了，抱歉

**Reply by Extra-Creativity** [reply]
我发现必须要extern "C"，否则函数还是要手动export，但是CMake好像没这个问题，这是什么原因？

**Reply by star-hengxing** [reply]
用不用 `extern "C"` 和导出符号无关，你可以对比 cmake 和 `xmake -v` 的编译命令

**Reply by Extra-Creativity** [reply]
我发现是`, {export_classes = true}`的问题，C++就算只export函数也要加这个选项，否则.def里就没有相应的符号；CMake没有这个问题。感觉这个`export_classes`的名字还挺迷惑的。

**Reply by star-hengxing** [reply]
@waruqi 自动导出符号的 rule 可以通过判断 target 有无 cpp 文件来自动开启导出，而不需要手动配置？

**Reply by waruqi** [reply]
不能，自己配置，自动导出会导出很多不需要的私有符号和接口，一般都是用户自己配置 def  或者走 代码显式导出的。。

这个 rule 只是没办法情况下的 偷懒导出方案，会导出很多不需要的私有符号和类，除非必要，一般是不用的。
